In [1]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
import os
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
ames_raw = pd.read_csv('Ames_HousePrice.csv', index_col=0)
print(ames_raw.shape)
print(ames_raw.columns)

(2580, 81)
Index(['PID', 'GrLivArea', 'SalePrice', 'MSSubClass', 'MSZoning',
       'LotFrontage', 'LotArea', 'Street', 'Alley', 'LotShape', 'LandContour',
       'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1',
       'Condition2', 'BldgType', 'HouseStyle', 'OverallQual', 'OverallCond',
       'YearBuilt', 'YearRemodAdd', 'RoofStyle', 'RoofMatl', 'Exterior1st',
       'Exterior2nd', 'MasVnrType', 'MasVnrArea', 'ExterQual', 'ExterCond',
       'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1',
       'BsmtFinSF1', 'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF',
       'Heating', 'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF',
       '2ndFlrSF', 'LowQualFinSF', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       

### I. Exploratory SLR
Run a simple linear regression on every column, giving birds-eye view of other variables' contributions to 'SalePrice'. Value Errors indicate categorical variables that will need dummifying/encoding later. Error for NaN's- indicate missing data

In [3]:
#Simple Linear Regression, fit simple linear model for every variable

#col = 'SaleCondition' #I manually looped these to catch errors individually as they came up.

#y = ames_raw[['SalePrice']]
#reg = LinearRegression()

#X = ames_raw[[col]]
    
#reg.fit(X, y)
#print(f" {col}: {reg.score(X, y)}")

PID: 0.05120099648251408  
GrLivArea: 0.5183732056699006  
MSSubClass: 0.007155008656742345  
MSZoning: VE  
'LotFrontage': NaN's  
'LotArea': .07531921127368835  
Street: VE  
Alley: VE  
'LotShape': VE  
'LandContour': VE  
'Utilities': VE  
'LotConfig': VE  
'LandSlope': VE  
'Neighborhood': VE  
'Condition1': VE  
'Condition2': VE  
'BldgType': VE  
'HouseStyle': VE  
OverallQual: 0.6248925373371652  
OverallCond: 0.010219487984851061  
YearBuilt: 0.2963386692622052  
YearRemodAdd: 0.2645135075446048  
'RoofStyle': VE  
'RoofMatl': VE  
'Exterior1st': VE  
'Exterior2nd': VE  
'MasVnrType': VE  
'ExterQual': VE  
'ExterCond': VE  
'Foundation': VE  
'BsmtQual': VE  
'BsmtCond': VE  
'BsmtExposure': VE  
'BsmtFinType1': VE  
'BsmtFinSF1': NaN's
'BsmtFinType2': VE  
'BsmtFinSF2': NaN's
'BsmtUnfSF': NaN's
'TotalBsmtSF': NaN's
'Heating': VE  
'HeatingQC': VE  
'CentralAir': VE  
'Electrical': VE  
1stFlrSF: 0.4129516427126516  
2ndFlrSF: 0.06924468617633406  
LowQualFinSF: 0.00095079438633483  
'BsmtFullBath': NaN's
'BsmtHalfBath': NaN's
FullBath: 0.28620549197746115
HalfBath: 0.08172451792098179
BedroomAbvGr: 0.02304512827536065
KitchenAbvGr: 0.01299070277321901
'KitchenQual': VE  
TotRmsAbvGrd: 0.24031508853002137
'Functional': VE  
Fireplaces: 0.23832326199307086
'FireplaceQu': VE  
'GarageType': VE  
'GarageYrBlt': NaN's
'GarageFinish': VE  
'GarageCars': NaN's
'GarageArea': NaN's
'GarageQual': VE
'GarageCond': VE
'PavedDrive': VE
WoodDeckSF: 0.11104333043218384
OpenPorchSF: 0.09396687512928903
EnclosedPorch: 0.015186832015373852
3SsnPorch: 0.0009771955140200905
ScreenPorch: 0.01475335746883455
PoolArea: 0.0009513780180608888
'PoolQC': VE
'Fence': VE
'MiscFeature': VE
MiscVal: 0.0003563384769366129
MoSold: 0.00015852081570277843
YrSold: 1.8925326150665e-06
'SaleType': VE
'SaleCondition': VE


#### Top Numerical R2's (no train-test or CV): 
OverallQual: 0.6248925373371652 -- I am gravely concerned about this column. 'Overall Quality' is feature engineering already built into the dataset- a subjective human metric that would be incredibly difficult to keep consistent and evaluate over time. Not to mention its distribution, which is heavily weighted so that the majority of the homes are 'above average'. I can't deny its predictive power, but it's fishy to me and I suspect it would receive a lot of scrutiny in a professional setting. I bristle at the presentations where students report that to make a house cost more, you have to increase square footage and its 'overall quality'- which feels like a cop-out.

GrLivArea (above grade living area with R^2 of .5183732056699006)  
1stFlrSF: 0.4129516427126516  
YearBuilt: 0.2963386692622052  
FullBath: 0.28620549197746115  
YearRemodAdd: 0.2645135075446048  

In [4]:
#Look at non-numeric columns- can drop in 'MSSubClass' too

y = ames_raw[['SalePrice']]
reg2 = LinearRegression()

for col in ames_raw.drop(columns = ['PID', 'SalePrice']).columns:

    if str(ames_raw[col].dtype) == 'object':

        x = ames_raw[[col]]
     
        x = pd.get_dummies(x, columns = [col], drop_first=True)
        
        x = x.replace({False: 0, True: 1})
        
        reg2.fit(x, y)
        
        print(f" {col}: {reg2.score(x, y)}")

    else:
        pass

 MSZoning: 0.11600712743510999
 Street: 0.0025456399344537806
 Alley: 1.1122790402051663e-06
 LotShape: 0.09415497298779363
 LandContour: 0.03234993583246448
 Utilities: 0.0011604270772369585
 LotConfig: 0.021107163245754457
 LandSlope: 0.0063239454783983184
 Neighborhood: 0.5598077663859118
 Condition1: 0.04531604555423885
 Condition2: 0.024720008929335457
 BldgType: 0.03303709816401801
 HouseStyle: 0.07103279739739266
 RoofStyle: 0.05870948148127286
 RoofMatl: 0.020763632343995697
 Exterior1st: 0.15513439019561037
 Exterior2nd: 0.15380427767670357
 MasVnrType: 0.17102556348836406
 ExterQual: 0.4853009289061254
 ExterCond: 0.021968757422164087
 Foundation: 0.26713508906430017
 BsmtQual: 0.3162074986324529
 BsmtCond: 0.04840595199785991
 BsmtExposure: 0.14879901006583107
 BsmtFinType1: 0.21719133838158222
 BsmtFinType2: 0.017428007577154014
 Heating: 0.007934972875623836
 HeatingQC: 0.20235536653584862
 CentralAir: 0.06642840369204306
 Electrical: 0.0564879727663965
 KitchenQual: 0.457

In [5]:
x = ames_raw[['MSSubClass']]
     
x = pd.get_dummies(x, columns = ['MSSubClass'], drop_first=False)
        
x = x.replace({False: 0, True: 1})
        
reg2.fit(x, y)
        
print(f" 'MSSubClass': {reg2.score(x, y)}")

 'MSSubClass': 0.2398439674330033


### Top Categorical R^2's (no train-test or CV): 
Neighborhood: 0.5597828887109129  
ExterQual: 0.48500289777659467  
BsmtQual: 0.4813818328427477  
KitchenQual: 0.45734015822742424  

### II. Impute Missing Data

## *Numerical Columns with Missing Values:*  
'LotFrontage'  
'BsmtFinSF1'   
'BsmtFinSF2'   
'BsmtUnfSF'   
'TotalBsmtSF'   
'BsmtFullBath'   
'BsmtHalfBath'   
'GarageYrBlt'   
'GarageCars'   
'GarageArea'  

### LotFrontage

In [6]:
#462 Rows with 'LotFrontage' NaNs

missingmask = ames_raw[ames_raw['LotFrontage'].isna()]
print(len(missingmask))

lot_traits = ['MSSubClass', 'MSZoning', 'LotArea', 'LotShape', 'LandContour', 'LotConfig']

for col in lot_traits:
    print(missingmask[col].value_counts(normalize=True))

#I don't see a compelling overall trend for buildings sold with missing 'LotFrontage' values.
#Missing Completely at Random? Is it possible to have a home without any lotfrontage?

462
MSSubClass
20     0.370130
60     0.277056
120    0.071429
80     0.064935
50     0.051948
90     0.036797
85     0.032468
160    0.028139
30     0.025974
70     0.021645
190    0.008658
75     0.004329
40     0.002165
180    0.002165
150    0.002165
Name: proportion, dtype: float64
MSZoning
RL         0.870130
RM         0.075758
FV         0.041126
RH         0.008658
I (all)    0.002165
C (all)    0.002165
Name: proportion, dtype: float64
LotArea
7500     0.010823
4426     0.008658
11000    0.008658
11250    0.008658
7750     0.008658
           ...   
11677    0.002165
6120     0.002165
12099    0.002165
11885    0.002165
8826     0.002165
Name: proportion, Length: 414, dtype: float64
LotShape
IR1    0.653680
Reg    0.277056
IR2    0.058442
IR3    0.010823
Name: proportion, dtype: float64
LandContour
Lvl    0.891775
Low    0.051948
Bnk    0.032468
HLS    0.023810
Name: proportion, dtype: float64
LotConfig
Inside     0.564935
Corner     0.205628
CulDSac    0.179654
FR2        0.

In [13]:
pd.set_option('display.max_columns', None)

neighborhood_mean_frontage = ames_raw.groupby('Neighborhood')['LotFrontage'].mean()

ames_raw['LotFrontage'] = ames_raw.apply(
    lambda row: neighborhood_mean_frontage[row['Neighborhood']] if\
    pd.isna(row['LotFrontage']) else row['LotFrontage'], axis = 1)
    
missingmask = ames_raw[ames_raw['LotFrontage'].isna()]
print(missingmask)

#These last remaining values are missing because there is no lot frontage recorded
#in these neighborhoods for us to take the average frontage!

#Impute the last three with mean
ames_raw['LotFrontage'] = ames_raw['LotFrontage'].fillna(ames_raw['LotFrontage'].mean())

           PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
110  916253320       1502     330000         120       RM          NaN   
684  916252170       1295     230000         120       RM          NaN   
205  907230240       1320     137000         160       RH          NaN   

     LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
110     9763   Pave   NaN      IR1         Lvl    AllPub    Inside       Gtl   
684     8239   Pave   NaN      IR1         Lvl    AllPub    Inside       Gtl   
205     3612   Pave   NaN      IR1         Lvl    AllPub    Inside       Gtl   

    Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  \
110      GrnHill       Norm       Norm   TwnhsE     1Story            7   
684      GrnHill       Norm       Norm   TwnhsE     1Story            7   
205      Landmrk       Norm       Norm    Twnhs     2Story            6   

     OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st  \
110  

In [14]:
y = ames_raw[['SalePrice']]
reg = LinearRegression()

X = ames_raw[['LotFrontage']]
    
reg.fit(X, y)
print(f" 'LotFrontage': {reg.score(X, y)}")
#imputing by neighborhood instead of imputing across the whole dataset brought the R^2 up\
#from .114 to .124 :\

 'LotFrontage': 0.12408330972516857


## 'Basement Square Footage'

In [15]:
mask = ames_raw['BsmtFinSF1'].isna()

ames_raw.loc[mask, ['BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF']] = (0, 0, 0, 0)

#Observations with Nans in 'BsmtFinSF1' don't have basements: Impute with 0's

In [16]:
y = ames_raw[['SalePrice']]
reg = LinearRegression()

for col in ['BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF']:
    X = ames_raw[[col]]
    
    reg.fit(X, y)
    print(f"{col}: {reg.score(X, y)}")

#TotalBsmtSF!

BsmtFinSF1: 0.21258570280531774
BsmtFinSF2: 0.0005138356647375053
BsmtUnfSF: 0.028244021234264394
TotalBsmtSF: 0.42574893740677966


## Basement Bathrooms

In [17]:
mask = ames_raw['BsmtFullBath'].isna()

ames_raw[mask]
ames_raw.loc[mask, ['BsmtFullBath', 'BsmtHalfBath']] = (0, 0)

#NaN's for Full Bathrooms in Basement? It's because they don't have a basement. Impute 0's

In [18]:
y = ames_raw[['SalePrice']]
reg = LinearRegression()

for col in ['BsmtFullBath', 'BsmtHalfBath']:
    X = ames_raw[[col]]
    
    reg.fit(X, y)
    print(f"{col}: {reg.score(X, y)}")

BsmtFullBath: 0.08453860463050078
BsmtHalfBath: 0.0011414971693050058


## Garage
'GarageYrBlt', 'GarageCars', 'GarageArea'

In [19]:
missingmask = ames_raw[ames_raw['GarageYrBlt'].isna()]
missingmask[missingmask['GarageType'] != 'No Garage']
mask = (ames_raw['PID'] == 910201180)

ames_raw.loc[mask, ['GarageYrBlt', 'GarageArea']] = (ames_raw['GarageYrBlt'].mean(), ames_raw['GarageArea'].mean())

mask = (ames_raw['PID'] == 903426160)

ames_raw.loc[mask, 'GarageYrBlt'] = (ames_raw['GarageYrBlt'].mean())
#All remaining 'GarageYrBlt' NaNs are for houses without garages. Impute with 0 to retain data type of column
#ames_raw['GarageYrBlt'].isna()
ames_raw['GarageYrBlt'] = ames_raw['GarageYrBlt'].fillna(0)

In [20]:
ames_raw['GarageCars'] = ames_raw['GarageCars'].fillna(value = 0.0)

In [21]:
y = ames_raw[['SalePrice']]
reg = LinearRegression()

for col in ['GarageYrBlt', 'GarageCars', 'GarageArea']:
    X = ames_raw[[col]]
    
    reg.fit(X, y)
    print(f"{col}: {reg.score(X, y)}")

GarageYrBlt: 0.06073326732327766
GarageCars: 0.4075857726993316
GarageArea: 0.40319406184284


## R^2 from Simple Linear Regression: Top Ten (No Train-Test or CV)

*1*. OverallQual: 0.6248925373371652  
*2*. Neighborhood: 0.5597828887109129  
*3*. GrLivArea: .5183732056699006  
*4*. ExterQual: 0.48500289777659467  
*5*. BsmtQual: 0.4813818328427477  
*6*. KitchenQual: 0.45734015822742424  
*7*. TotalBsmtSF: 0.42574893740677966  
*8*. 1stFlrSF: 0.4129516427126516  
*9*. GarageCars: 0.4075857726993316  
*10*.GarageArea: 0.40319406184284 (obviously highly correlated with GarageCars!)

## Categorical Columns with Missing Values

In [22]:
ames_raw.columns[ames_raw.isna().any()].tolist()

['Alley',
 'MasVnrType',
 'MasVnrArea',
 'BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'BsmtFinType1',
 'BsmtFinType2',
 'Electrical',
 'FireplaceQu',
 'GarageType',
 'GarageFinish',
 'GarageQual',
 'GarageCond',
 'PoolQC',
 'Fence',
 'MiscFeature']

In [23]:
#Alley
ames_raw['Alley'] = ames_raw['Alley'].fillna('No Alley')

In [24]:
#Filter to where 'MasVnrType' is null and 'MasVnrArea' is NOT null:
missingmask = ames_raw[ames_raw['MasVnrType'].isna()]
missingmask[pd.notna(missingmask['MasVnrArea']) & missingmask['MasVnrArea'] != 0]

PID  GrLivArea  SalePrice  MSSubClass MSZoning  LotFrontage  \
115  534129230       1790     165150          60       RL    80.000000   
604  528138010       1808     324000          20       RL   102.000000   
138  535106140        912     114500          20       RL    70.000000   
231  527451110       1365     125000         160       RM    24.000000   
177  527166010       1983     225000          60       RL    73.362637   
284  902427140        968     104500          20       RM    52.000000   
290  533352075       2620     190000          90       RL    75.388889   

     LotArea Street     Alley LotShape LandContour Utilities LotConfig  \
115    10400   Pave  No Alley      Reg         Lvl    AllPub    Inside   
604    13514   Pave  No Alley      IR1         Lvl    AllPub    Corner   
138    10150   Pave  No Alley      Reg         Lvl    AllPub    Inside   
231     2368   Pave  No Alley      Reg         Lvl    AllPub    Inside   
177    10762   Pave  No Alley      IR1         Lvl    AllPub   CulDSac   
284     8626   Pave  No Alley      Reg         Lvl    AllPub    Inside   
290    18890   Pave  No Alley      IR1         Lvl    AllPub    Inside   

    LandSlope Neighborhood Condition1 Condition2 BldgType HouseStyle  \
115       Gtl       NWAmes       Norm       Norm     1Fam     2Story   
604       Gtl      NridgHt       Norm       Norm     1Fam     1Story   
138       Gtl        NAmes      Feedr       Norm     1Fam     1Story   
231       Gtl       BrDale       Norm       Norm   TwnhsE     2Story   
177       Gtl      Gilbert       Norm       Norm     1Fam     2Story   
284       Gtl      OldTown       Norm       Norm     1Fam     1Story   
290       Gtl       Sawyer      Feedr       RRAe   Duplex     1.5Fin   

     OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl  \
115            6            5       1972          1972     Gable  CompShg   
604            9            5       2008          2008       Hip  CompShg   
138            5            5       1958          1958     Gable  CompShg   
231            5            6       1970          1970     Gable  CompShg   
177            7            5       1999          1999     Gable  CompShg   
284            4            6       1956          1956     Gable  CompShg   
290            5            5       1977          1977      Shed  CompShg   

    Exterior1st Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond  \
115     VinylSd     VinylSd        NaN       288.0        TA        TA   
604     VinylSd     VinylSd        NaN       285.0        Ex        TA   
138     Wd Sdng     Wd Sdng        NaN         1.0        TA        TA   
231     HdBoard     HdBoard        NaN       312.0        TA        TA   
177     VinylSd     VinylSd        NaN       344.0        Gd        TA   
284     MetalSd     MetalSd        NaN         1.0        TA        TA   
290     Plywood     Plywood        NaN         1.0        TA        TA   

    Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1  \
115     CBlock       TA       TA           No          Rec       247.0   
604      PConc       Ex       TA           No          GLQ      1142.0   
138     CBlock       TA       TA           No          Rec       456.0   
231     CBlock       TA       TA           No          LwQ       765.0   
177      PConc       Gd       TA           No          GLQ       694.0   
284     CBlock      NaN      NaN          NaN          NaN         0.0   
290     CBlock       Gd       TA           No          GLQ       498.0   

    BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC  \
115          Unf         0.0      485.0        732.0    GasA        Gd   
604          Unf         0.0      632.0       1774.0    GasA        Ex   
138          Unf         0.0      456.0        912.0    GasA        Ex   
231          Unf         0.0        0.0        765.0    GasA        TA   
177          Unf         0.0      284.0        978.0    GasA        Ex   
2

In [25]:
#Where masonry is 1 sq ft? Remove, replace with 0
mask1 = (ames_raw['MasVnrType'].isna()) & (ames_raw['MasVnrArea'].notna()) & (ames_raw['MasVnrArea'] == 1)
ames_raw.loc[mask1, ['MasVnrType', 'MasVnrArea']] = ('No MasVnr', 0)

In [26]:
#Where masonry is more than 1 sq ft? Retain, Impute MasVnrType with mode: 'BrkFace'
mask2 = (ames_raw['MasVnrType'].isna()) & (ames_raw['MasVnrArea'].notna()) & (ames_raw['MasVnrArea'] != 0)
ames_raw.loc[mask2, 'MasVnrType'] = 'BrkFace'

In [27]:
ames_raw['MasVnrType'] = ames_raw['MasVnrType'].fillna(value = 'No MasVnr')

In [28]:
#MasVnrArea
missingmask = ames_raw[ames_raw['MasVnrArea'].isna()]
missingmask
#Masonry Veneer Area is Nan in these columns because 'MasVnrType' is exclusively 'None': Impute 0's

ames_raw['MasVnrArea'] = ames_raw['MasVnrArea'].fillna(0)

### Basement Values

In [29]:
mask = (ames_raw['BsmtQual'].isna() & ames_raw['BsmtFinSF1'].notna())
ames_raw.loc[mask, ['BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2']] = ('No Bsmt', 'No Bsmt',\
                                                                                              'No Bsmt', 'No Bsmt', 'No Bsmt')

In [30]:
missingmask = ames_raw['BsmtQual'].isna()

ames_raw.loc[missingmask, ['BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2']] = ('No Bsmt', 'No Bsmt',\
                                                                                              'No Bsmt', 'No Bsmt', 'No Bsmt')

In [31]:
mask = ames_raw['BsmtExposure'].isna()
ames_raw.loc[mask, 'BsmtExposure'] = 'No'

In [32]:
mask = ames_raw['BsmtFinType2'].isna()
ames_raw.loc[mask, ['BsmtFinType2', 'BsmtUnfSF', 'TotalBsmtSF']] = ('Unf', 479.0, 1603.0)

### Electrical

In [33]:
ames_raw['Electrical'] = ames_raw['Electrical'].fillna(value = 'SBrkr')

### Fireplace

In [34]:
mask = ames_raw['FireplaceQu'].isna()
ames_raw.loc[mask, 'FireplaceQu'] = 'No Fireplace'

### Garage

In [35]:
missingmask = ames_raw['GarageType'].isna()
ames_raw.loc[missingmask, ['GarageType', 'GarageFinish', 'GarageQual', 'GarageCond']] =\
    ('No Garage', 'No Garage', 'No Garage', 'No Garage')
#Where GarageType is NaN, all other Garage Attributes are 0 or NaN

In [36]:
mask = (ames_raw['PID'] == 910201180)

ames_raw.loc[mask, ['GarageFinish', 'GarageCars', 'GarageQual', 'GarageCond']] =\
    ('Unf', 1, 'TA', 'TA')

In [37]:
mask = (ames_raw['PID'] == 903426160)

ames_raw.loc[mask, ['GarageFinish', 'GarageQual', 'GarageCond']] =\
    ('Unf', 'TA', 'TA')

### PoolQC, MiscFeature, Fence

In [38]:
mask = ames_raw['PoolQC'].isna() & ames_raw['PoolArea'].notna() & (ames_raw['PoolArea'] != 0)
ames_raw[mask]
#If PoolQC is missing, the square footage is listed as 0, safe to impute with 'None'

ames_raw['PoolQC'] = ames_raw['PoolQC'].fillna('No Pool')

In [39]:
mask = ames_raw['MiscFeature'].isna() & ames_raw['MiscVal'].notna() & (ames_raw['MiscVal'] != 0)
ames_raw[mask]

#If MiscFeature is missing, there are no dangling 'MiscVal' values to complicate imputing with 'None'

ames_raw['MiscFeature'] = ames_raw['MiscFeature'].fillna('No Misc')

In [40]:
ames_raw['Fence'] = ames_raw['Fence'].fillna('No Fence')

In [41]:
ames = ames_raw.copy()

### III. Feature Engineering

In [42]:
#Total Bathrooms
ames['TotalBath'] = ames['BsmtFullBath'] + ames['FullBath'] +\
                       (ames.BsmtHalfBath * .5) + (ames.HalfBath * .5)
#Total Square Footage: Above Grade + Basement
ames['TotalSF'] = ames['GrLivArea'] + ames['TotalBsmtSF']

#High Quality SF
ames['HQSF'] = ames['TotalSF'] - ames['LowQualFinSF']

#Years Between Remodel and Build
ames['yrsbtwn'] = ames['YearRemodAdd'] - ames['YearBuilt']

#Bedrooms x Total Bathrooms
ames['BedxBath'] = ames['TotalBath'] * ames['BedroomAbvGr']

#Rooms x Total Bathrooms
ames['RoomsxBath'] = ames['TotalBath'] * ames['TotRmsAbvGrd']

#Any Fireplaces
ames['FireplaceYN'] = ames['Fireplaces'].apply(lambda x: 1 if x != 0 else 0)

#Combine Porches
ames['TotalPorchSF'] = ames[['WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch',\
                                    '3SsnPorch', 'ScreenPorch']].sum(axis=1)
#Change Type
ames['MSSubClass'] = ames['MSSubClass'].astype(str)

### IV. Missing Data Imputed + Feature Engineering Done: Export as 'ames.csv'

In [44]:
current_directory = os.getcwd()

file_path = os.path.join(current_directory, "ames.csv")

ames.to_csv(file_path)